Subject: [PATCH] xtables-addons: adapt RTSP helper API for Linux 7.3

Adapt RTSP connection tracking helper registration to the Linux 7.3
API changes, including the protocol-field updates. Retain registered
helper pointers so cleanup after partial registration failures
unregisters only successfully registered helpers.

--- a/extensions/rtsp/nf_conntrack_rtsp.c
+++ b/extensions/rtsp/nf_conntrack_rtsp.c
@@ -676,6 +676,9 @@
 }
 
 static struct nf_conntrack_helper rtsp_helpers[MAX_PORTS];
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(7,2,0)
+static struct nf_conntrack_helper *rtsp_helper_ptrs[MAX_PORTS];
+#endif
 static char rtsp_names[MAX_PORTS][10];
 
 /* This function is intentionally _NOT_ defined as __exit */
@@ -685,7 +688,12 @@
 	int i;
 	for (i = 0; i < num_ports; i++) {
 		pr_debug("unregistering port %d\n", ports[i]);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(7,2,0)
+		if (rtsp_helper_ptrs[i])
+			nf_conntrack_helper_unregister(rtsp_helper_ptrs[i]);
+#else
 		nf_conntrack_helper_unregister(&rtsp_helpers[i]);
+#endif
 	}
 	kfree(rtsp_buffer);
 }
@@ -720,10 +728,19 @@
 	for (i = 0; (i < MAX_PORTS) && ports[i]; i++) {
 		hlpr = &rtsp_helpers[i];
 		memset(hlpr, 0, sizeof(struct nf_conntrack_helper));
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(7,3,0)
+		hlpr->nfproto = NFPROTO_IPV4;
+		hlpr->l4proto = IPPROTO_TCP;
+#else
 		hlpr->tuple.src.l3num = AF_INET;
 		hlpr->tuple.src.u.tcp.port = htons(ports[i]);
 		hlpr->tuple.dst.protonum = IPPROTO_TCP;
+#endif
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(7,2,0)
+		hlpr->expect_policy[0] = rtsp_exp_policy;
+#else
 		hlpr->expect_policy = &rtsp_exp_policy;
+#endif
 		hlpr->me = THIS_MODULE;
 		hlpr->help = help;
 
@@ -743,7 +760,11 @@
 #endif
 		pr_debug("port #%d: %d\n", i, ports[i]);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(7,2,0)
+		ret = nf_conntrack_helper_register(hlpr, &rtsp_helper_ptrs[i]);
+#else
 		ret = nf_conntrack_helper_register(hlpr);
+#endif
 
 		if (ret) {
 			printk("nf_conntrack_rtsp: ERROR registering port %d\n", ports[i]);
